oeps/oep-0007.rst

```diff
@@ -0,0 +1,462 @@
+
+Other problems
+--------------
+
+If you find other incompatibilities, a shim will likely be found in [the `six`
+library](https://pythonhosted.org/six/).  For incompatibilies with no other
+solution, edX will maintain a repository of compatibility shims (edx-compat?).
+Ideally, all edx-maintained code that implements different behavior based on
+Python version will be in this repo.
+
+When writing code that explicitly switches based on version, do
+
+    if six.PY2:  # or future.PY2
+        do_python2_thing()
+    else:
+        do_python3_thing()
+
+Do not explicitly call out `six.PY3` or `future.PY3`.  This should be more
+future-compatible with a potential future Python 4.
+
+
+Changing Practice
++++++++++++++++++
+
+Changing internal code practices to ease conversion will require a
+three-pronged approach of documentation, socialization, and tooling.  To start,
+we need to update the official edX code style guide to mandate compatible code
+practices.  To socialize these practices among our engineers, we will announce
+our efforts to migrate to Python 3 during an engineering all-hands meeting,
+offer a workshop in writing compatible code, and promote awareness of
+incompatibilities during code reviews.  Additionally, we will host regular
+Python 3 office hours to help answer questions and troubleshoot problems that
+arise during migration.
+
+Appropriate tooling will help.  Tests should be configured to run under both
+Python 2 or 3 (tox helps with this).  As it would double the cost of running
+tests to run both all the time, we may only want to enable Python 3 tests in
+the CI environment for repos that are actively being migrated.  We still want
+to ensure that Python 3 compatibility is improving on other projects, so a
+lightweight metric that can be run every time would also be useful, such as
+running pylint with the `caniusepython3.pylint_checker` extension.  Making these checks
+mandatory in a similar way to our current quality will ensure that
+compatibility is improving (TODO: timeline for this?).
+
+
+Migrating projects
+++++++++++++++++++
+
+We should be able to migrate individual applications to Python 3 independently.
+To begin with, we should pilot the process using a relatively small (but
+complex enough to provide useful information) IDA.  As we go, we will document
+the process, find pain points, figure out ways of dealing with them, and
+continue to improve our process.
+
+
+Order of migrations
++++++++++++++++++++
+
+* IDAs that we want to continue supporting in the future
+  * Old IDAs (that we want to replace) should not be upgraded, but we will need
+    to prioritize replacement to occur during the migration timeframe.
+* Implement remote execution of xBlocks (to allow a window of bicompatibility
+  for external xblocks)
+* edx-platform
+  * Deploy xblocks separately to test remote execution
+  * Migrate to Python 3
+  * Upgrade external xblocks as needed, and support partners who wish to do the
+    same
```

Comment:
This isn't valid rst formatting for nested lists; the type of bullet needs to change for the 2nd level